### PR TITLE
Add option to throw on encountering direct eval

### PR DESCRIPTION
**Pull request by bakkot**
I'm not thrilled about throwing a RuntimeException. Is there a better kind of thing to throw? Should we make one?

### CODE REVIEW

File: src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java [hunk 1]
```diff
@@ -119,77 +120,56 @@
...
-		this.isCandidateForInlining = list -> false;
-	}
-
-	Explicator(@Nonnull Script script, @Nonnull F<ImmutableList<Directive>, Boolean> isCandidateForInlining) {
-		this.program = Either.left(script);
-		this.scope = ScopeAnalyzer.analyze(script);
-		this.jumpMap = FinallyJumpReducer.analyze(script);
+	final boolean throwOnEval;
+
+	Explicator(@Nonnull Program program, @Nonnull F<ImmutableList<Directive>, Boolean> isCandidateForInlining, boolean throwOnEval) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java]: Let's have additional constructors that give defaults for the recently-added parameters.

File: src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java [hunk 2]
```diff
@@ -796,8 +776,10 @@ NodeWithValue explicateExpression(Expression expression, boolean keepValue, bool
 				this.currentState.exitInlineFunction();
 				return functionBlock;
 			}
-			// abort on direct eval. todo maybe warn.
 			if (c.callee instanceof IdentifierExpression && (((IdentifierExpression) c.callee).name.equals("eval"))) {
+				if (this.throwOnEval) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java]: How about instead we have them pass in a function that returns the list of instructions to use for a direct eval? Default to the list containing just `Halt`, but this also allows them to throw a `RuntimeException`.

**Review by michaelficarra — APPROVED** [review]

File: src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java (hunk 1 above)
Comment (bakkot) [reply src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java]: The constructor is not public. There's overloads for the public static methods; that seems sufficient.

File: src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java (hunk 2 above)
Comment (bakkot) [reply src/main/java/com/shapesecurity/shift/es2016/semantics/Explicator.java]: Done.

**Comment by michaelficarra** [comment]
@bakkot The supplier should have produced a list of nodes, not a single node. A user may want to supply the empty list, for instance.

**Comment by bakkot** [comment]
@michaelficarra That would not work with the type system. A call expression compiles to a `NodeWithValue`, since it can appear in for example `x = eval()`.

The user can of course choose to provide the node whose value is `undefined`, or a `BlockWithValue` which contains more complex logic.

**Comment by michaelficarra** [comment]
Okay I am satisfied with `BlockWithValue`.

**Pull request merged**